text/3806-align-attr.md

```diff
@@ -0,0 +1,380 @@
+possible. One way of doing this is to force the alignment of the value to be at
+least the size of the cache line, or perhaps the greatest common denominator of
+the value and cache line sizes.
+
+The simplest way of accomplishing this in Rust today is to use a wrapper struct
+with a `#[repr(align(…))]` attribute:
+
+```rust
+type SomeLargeType = [[u8; 64]; 21];
+
+#[repr(align(128))]
+struct CacheAligned<T>(T);
+
+static LOOKUP_TABLE: CacheAligned<SomeLargeType> = CacheAligned(SomeLargeType {
+    data: todo!(),
+});
+```
+
+However, this approach has several downsides:
+
+- It requires defining a separate wrapper type.
+- It changes the type of the item, which may not be allowed if it is part of the
+  crate's public API.
+- It may add padding to the value, which might not be necessary or desirable.
+
+In some cases, it can also improve performance to align function items in the
+same way.
+
+## Required alignment for low-level use cases
+
+Some low-level use-cases (for example, the [RISC-V `mtvec`
+register](https://five-embeddev.com/riscv-priv-isa-manual/Priv-v1.12/machine.html#machine-trap-vector-base-address-register-mtvec))
+require functions or statics to have a certain minimum alignment.
+
+## Interoperating with systems that have types where size is not a multiple of alignment
+
+In Rust, a type’s size is always a multiple of its alignment. However, there are
+other languages that can interoperate with Rust, where this is not the case
+(WGSL, for example). It’s important for Rust to be able to represent such
+structures.
+
+# Explanation
+[explanation]: #explanation
+
+The `align` attribute is a new inert, built-in attribute that can be applied to
+ADT fields, `static` items, function items, and local variable declarations. The
+attribute accepts a single required parameter, which must be a power-of-2
+integer literal from 1 up to 2<sup>29</sup>. (This is the same as
+`#[repr(align(…))]`.)
+
+Multiple `align` attributes may be present on the same item; the highest
+alignment among them will be used. The compiler may signal this case with a
+warn-by-default lint.
+
+## On ADT fields
+
+The `align` attribute may be applied to any field of any `struct`, `enum`, or
+`union` that is not `#[repr(transparent)]`.
+
+```rust
+#[repr(C)]
+struct Foo {
+    #[align(8)]
+    a: u32,
+}
+
+enum Bar {
+    Variant(#[align(16)] u128),
+}
+
+union Baz {
+    #[align(16)]
+    a: u32,
+}
+```
+
+The effect of the attribute is to force the address of the field to have at
+least the specified alignment. If the field already has at least that
+alignment, due to the required alignment of its type or to a `repr` attribute on
+the containing type, the attribute has no effect.
+
+In contrast to a `repr(align(…))` wrapper struct, an `align` annotation does *not*
+necessarily add extra padding to force the field to have a size that is a
+multiple of its alignment. (The size of the containing ADT must still be a
+multiple of its alignment; that hasn't changed.)
```

Comment:
A field with `#[align(N)]` does raise the alignment of the containing ADT to be (at least) N, with all the potential padding that entails, right? I don't see how this can work otherwise, but I haven't seen it spelled out explicitly.

Reply:
Correct.